File: GFG/Reverse Spiral Form of Matrix/Reverse Spiral Form of Matrix.cpp
class Solution {
  public:
    vector<int> reverseSpiral(int R, int C, vector<vector<int>>&a)  {
        // code here
        
          vector<int>ans;

       

        int top=0,left=0,right=C-1,bottom=R-1;

        while(top<=bottom and left<=right)

        { // top row

         if(top<=bottom)

            {

                for(int i=left;i<=right;i++)

                 ans.push_back(a[top][i]);

            

                 top++;

            }

            // right column

            if(left<=right){

            for(int i=top;i<=bottom;i++)

            {

                ans.push_back(a[i][right]);

            }

            right--;

            }

          //bottom row (reverse order)

          if(top<=bottom)

          {

              for(int i=right;i>=left;i--)

              {

                  ans.push_back(a[bottom][i]);

              }

              bottom--;

          }

          //left column (reverse order)

           if(left<=right)

          {

              for(int i=bottom;i>=top;i--)

              {

                  ans.push_back(a[i][left]);

              }

              left++;

          }

        }

        // reverse ans vector

      reverse(ans.begin(),ans.end());

        return ans;
    }
};

File: GFG/Reverse Spiral Form of Matrix/Reverse Spiral Form of Matrix.java
class Solution
{
    public int[] reverseSpiral(int R, int C, int[][] a)
    {
        // code here
        int[] ans=new int[R*C];
        int idx=R*C-1;
        
        int up=0;
        int down=a.length-1;
        int left=0;
        int right=a[0].length-1;
        
        
        
        while(idx>=0){
            //upper row
            for(int i=left;i<=right && idx>=0;i++){
                ans[idx--]=a[up][i];
            }
            up++;
            
            //last col
            for(int i=up;i<=down && idx>=0;i++){
                ans[idx--]=a[i][right];
            }
            right--;
            
            //lower row
            for(int i=right;i>=left && idx>=0;i--){
                ans[idx--]=a[down][i];
            }
            down--;
            
            for(int i=down;i>=up && idx>=0;i--){
                ans[idx--]=a[i][left];
            }
            left++;
        }
        
        return ans;
        
    }
}
